server/sonar-server/src/main/java/org/sonar/server/user/UserUpdater.java

```diff
@@ -199,45 +178,80 @@ private UserDto createNewUserDto(DbSession dbSession, NewUser newUser) {
     String email = updateUser.email();
-    if (updateUser.isEmailChanged() && validateEmailFormat(email, messages)) {
+    if (updateUser.isEmailChanged() && validateEmailFormat(email, messages) && !Objects.equals(userDto.getEmail(), email)) {
       userDto.setEmail(email);
+      return true;
     }
+    return false;
+  }
 
-    if (updateUser.isExternalIdentityChanged()) {
-      setExternalIdentity(userDto, updateUser.externalIdentity());
+  private static boolean updateExternalIdentity(UpdateUser updateUser, UserDto userDto) {
+    ExternalIdentity externalIdentity = updateUser.externalIdentity();
+    if (updateUser.isExternalIdentityChanged() && !isSameExternalIdentity(userDto, externalIdentity)) {
+      setExternalIdentity(userDto, externalIdentity);
       userDto.setSalt(null);
       userDto.setCryptedPassword(null);
-    } else {
-      String password = updateUser.password();
-      if (updateUser.isPasswordChanged() && validatePasswords(password, messages) && checkPasswordChangeAllowed(userDto, messages)) {
-        setEncryptedPassWord(password, userDto);
-      }
+      return true;
     }
+    return false;
+  }
 
-    if (updateUser.isScmAccountsChanged()) {
-      List<String> scmAccounts = sanitizeScmAccounts(updateUser.scmAccounts());
-      if (scmAccounts != null && !scmAccounts.isEmpty()) {
+  private static boolean updatePassword(UpdateUser updateUser, UserDto userDto, List<Message> messages) {
+    String password = updateUser.password();
+    if (!updateUser.isExternalIdentityChanged() && updateUser.isPasswordChanged() && validatePasswords(password, messages) && checkPasswordChangeAllowed(userDto, messages)) {
+      setEncryptedPassWord(password, userDto);
+      return true;
+    }
+    return false;
+  }
+
+  private boolean updateScmAccounts(DbSession dbSession, UpdateUser updateUser, UserDto userDto, List<Message> messages) {
+    String email = updateUser.email();
+    List<String> scmAccounts = sanitizeScmAccounts(updateUser.scmAccounts());
+    List<String> existingScmAccounts = userDto.getScmAccountsAsList();
+    if (updateUser.isScmAccountsChanged() && !(existingScmAccounts.containsAll(scmAccounts) && scmAccounts.containsAll(existingScmAccounts))) {
+      if (!scmAccounts.isEmpty()) {
         String newOrOldEmail = email != null ? email : userDto.getEmail();
         if (validateScmAccounts(dbSession, scmAccounts, userDto.getLogin(), newOrOldEmail, userDto, messages)) {
           userDto.setScmAccounts(scmAccounts);
         }
       } else {
         userDto.setScmAccounts((String) null);
       }
+      return true;
     }
+    return false;
+  }
 
-    if (!messages.isEmpty()) {
-      throw new BadRequestException(messages);
-    }
+  private static boolean isSameExternalIdentity(UserDto dto, @Nullable ExternalIdentity externalIdentity) {
+    return (externalIdentity == null && dto.getExternalIdentity() == null) ||
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png 'Severity: CRITICAL') Reduce the number of conditional operators (4) used in the expression (maximum allowed 3). [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1067)